Expensify/App
# [No QA] Use full Git history to populate StagingDeployCash

### Details
The main motivation behind this pull request is to use the full Git / GitHub history to populate the StagingDeployCash, rather than relying on one-off changes made in actions. This makes the whole process more robust: any time the StagingDeployCash is updated, it should have all the correct pull requests and deploy blockers. That means if there is a GH outage, then when the service is restored the StagingDeployCash should self-resolve to its correct state.

Here's a more detailed list of the changes made in this PR (in broad strokes)

1. Changes `GitUtils.getPullRequestsMergedBetween` to be synchronous instead of async. Before it was unnecessarily async, and an arbitrary source of extra complexity.
1. Removes `createNewStagingDeployCash` and `updateStagingDeployCash` from the `GithubUtils` class. This is now handled entirely in the `createOrUpdateStagingDeploy` GitHub Action (located at `.github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js`). Accordingly, I stripped some tests from `GitHubUtilsTest` and created a new automated test for the `createOrUpdateStagingDeploy` GH Action.
1. Removes the `NEW_PULL_REQUESTS` and `NEW_DEPLOY_BLOCKERS` inputs from the `createOrUpdateStagingDeploy` action. Instead, the action finds pull requests by looking at the tag on the previous StagingDeployCash, and parsing the git history between that tag and the latest tag to determine which PRs were merged since the last prod deploy. It finds deploy blockers by just querying the GH api for any open issues with the `DeployBlockerCash` label.
1. Updates the YML workflows to remove those unused inputs from usages of the `createOrUpdateStagingDeploy` action.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/164517

### Tests
1. Merge this pull request.
1. Verify that a staging deploy happens and that the `StagingDeployCash` is updated with this pull request and the new `BUILD` version.
1. Check off this pull request on the `StagingDeployCash`
1. Lock the `StagingDeployCash`, then immediately after merge a pull request.
1. Verify that the `StagingDeployCash` is updated to include that pull request.
1. Merge another pull request.
1. Verify that the the pull request gets the "deploy deferred" comment.
1. Verify that the pull request _is not_ deployed to staging and _is not_ added to the `StagingDeployCash`
1. Create a new issue labelled `DeployBlockerCash`
1. Verify that the `StagingDeployCash` is updated to include the new deploy blocker
1. Label this pull request with `DeployBlockerCash`
1. Verify that the `StagingDeployCash` is updated to list this pull request as a deploy blocker.
1. Unlock and relock the `StagingDeployCash`.
1. Verify that the `StagingDeployCash` is update to include the deferred pull request.
1. Verify that the items on the `StagingDeployCash` that were checked off remain checked off.
1. Merge another pull request. Verify that it gets the "deploy deferred" comment and is not deployed to staging / added to the `StagingDeployCash`.
1. Close the `StagingDeployCash` to trigger a production deploy. Verify that the prod deploy begins.
1. Verify that a new `StagingDeployCash` is created, including the second deferred PR.
1. Once the prod deploy completes, verify that every PR on the previous `StagingDeployCash` gets the "deployed to prod" comment.
